회원가입 모달 창 제작

jinhyukoo opened the pull request:
## 📑 제목

회원가입 모달 창 제작을 완료했습니다.

![녹화_2021_04_10_14_11_16_938](https://user-images.githubusercontent.com/46101366/114259130-b4e01e00-9a06-11eb-8720-c15717358a56.gif)

useDebounce hook은 다음 사이트를 참고했습니다.
- https://dev.to/gabe_ragland/debouncing-with-react-hooks-jci

대충 설명하면 다음과 같습니다.
1. 사용하고자 하는 컴포넌트에서 useState를 사용하여 상태를 선언한 후 이 상태값과 원하는 시간값을 useDebounce hook에 parameter로 전달해줍니다.
2. useDebounce 코드를 보시면 아시겠지만 useEffect를 활용하여 컴포넌트가 렌더링 되면 setTimeout이 시작되고 사라질 때 clearTimeout이 됩니다. 이를 토대로 정해진 시간만큼의 시간이 흐르면 정상적으로 debouncedValue가 리턴이 되고 그렇지 않으면 기존값이 리턴됩니다.
3. 정상적으로 바뀐값이 리턴된 경우 컴포넌트의 useEffect에서 바뀐 값을 인지하고 새롭게 렌더링을 함과 동시에 원하는 로직을 실행시킬 수 있습니다.

## 💬 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] 회원가입 모달 뷰 제작
- [x] debounce 적용, useDebounce 제작
- [x] 로그인 모달 뷰 예외 처리 적용


## 🚧 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점
- 현재 ModalContent에서 상태값들을 관리하고 있기 때문에 하나의 TextField를 사용하면 모든 Content들이 리렌더링 되는 상황입니다. 성능에 크게 영향을 끼치진 않으나 아주 조금이라도 최적화를 하려면 상태값을 가진 컴포넌트들로(TextField를 하나씩 만든다던가...) 또다시 분리해야 하나 하는 생각을 해봤습니다. 근데 이건 좀 아닌 것 같긴 합니다 ㅎ...

wooojini committed:
test: 스토리북 css reset 설정 추가

- 스토리북 테스팅시 스타일을 reset하고 개발할 수 있도록 reset.css import
  구문 추가

wooojini committed:
feat: SelectMenu 컴포넌트 제작

- SelectMenu 컴포넌트 제작 및 스토리 작성

jinhyukoo committed:
feat: SignUpModalContent 컴포넌트 기초 작업

jinhyukoo committed:
feat: SignUpModalContent 컴포넌트 기초 작업

jinhyukoo committed:
feat: SignUpModalPopup 컴포넌트 제작

jinhyukoo committed:
feat: SignUpModalPopup 컴포넌트 제작

jinhyukoo committed:
feat: ModalPopup 컴포넌트에 SignUp 추가

jinhyukoo committed:
feat: ModalPopup 컴포넌트에 SignUp 추가

jinhyukoo committed:
feat: SignUpModal 추가

- 텍스트 클릭시 Modal 보여짐
- UI 수정 (hover시 cursor 변경)
- props 수정

jinhyukoo committed:
feat: SignUpModal 추가

- 텍스트 클릭시 Modal 보여짐
- UI 수정 (hover시 cursor 변경)
- props 수정

wooojini committed:
test: 불필요한 테스트 파일 삭제

- 프로젝트 초기 테스트 환경 구성을 위한 샘플 테스트 코드 삭제

wooojini committed:
feat: range 유틸 함수 작성

- start~end 사이의 수를 offset 간격으로 반복하는 이터레이터 유틸 함수
- range 유틸 테스트 코드 작성

wooojini committed:
chore: eslint ++연산자 제한 옵션 off

- ++연산자 사용을 제한하는 옵션 off

wooojini committed:
style: Button 컴포넌트 스타일 수정

- Add Button px단위 rem단위로 수정

wooojini committed:
refactor: SelectMenu 컴포넌트 디자인 수정

- SelectMenu 컴포넌트 Popover 컴포넌트를 이용해서 직접 제작
- 외부에서 width값을 설정해서 크기를 조절하도록 수정 및 스토리북 코드
  반영

wooojini committed:
feat: TimeSelectMenu 컴포넌트 제작

- 시간 선택을 위한 TimeSelectMenu 컴포넌트 제작
- 기존 input type=time으로 제공되는 UI는 브라우저마다 다르고, 스타일
  커스텀에 한계가 존재해서 별도로 제작
- TimeSelectMenu 컴포넌트 스토리 작성

wooojini committed:
feat: LectureSearchFilterMenu 컴포넌트 제작

- 강의 검색 필터를 위한 LectureSearchFilterMenu 컴포넌트 제작
- LectureSearchFilterMenu 스토리 작성

wooojini committed:
feat: TimeTableAddForm 컴포넌트 제작

- 나만의 시간표 추가를 위한 TimeTableAddFrom 컴포넌트 제작
- TimeTableAddForm 스토리 작성

wooojini committed:
style: 메인페이지 레이아웃 수정

- LectureSearchFilterMenu, TimeTableAddForm 컴포넌트 메인페이지 레이아웃
  추가

jinhyukoo committed:
feat: useDebounce Hooks 제작

- 컴포넌트로부터 상태값을 받아와 일정 시간이 지난 후 해당 값을 리턴
- 시간이 지나기 전에 상태가 변하게 되면 컴포넌트가 리렌더링 되며 clearTimeout 실행
- hook을 가져다 쓰는 컴포넌트 입장에서는 debouncedValue의 존재여부로 판단 가능

jinhyukoo committed:
feat: useDebounce Hooks 제작

- 컴포넌트로부터 상태값을 받아와 일정 시간이 지난 후 해당 값을 리턴
- 시간이 지나기 전에 상태가 변하게 되면 컴포넌트가 리렌더링 되며 clearTimeout 실행
- hook을 가져다 쓰는 컴포넌트 입장에서는 debouncedValue의 존재여부로 판단 가능

jinhyukoo committed:
feat: Email, Password 유효성 검사 로직 추가

- useDebounce hook 사용

jinhyukoo committed:
feat: Email, Password 유효성 검사 로직 추가

- useDebounce hook 사용

jinhyukoo committed:
feat: Login 화면 유효성 검사 로직  추가

jinhyukoo committed:
feat: Login 화면 유효성 검사 로직  추가

jinhyukoo committed:
feat: autoComplete off 추가

- 사용자가 원치 않는 자동완성 방지

jinhyukoo committed:
feat: autoComplete off 추가

- 사용자가 원치 않는 자동완성 방지

jinhyukoo committed:
refactor: console.log 제거

jinhyukoo committed:
refactor: console.log 제거

jinhyukoo committed:
feat: SignUpModalContent 스토리북 추가

jinhyukoo committed:
feat: SignUpModalContent 스토리북 추가

wooojini commented on client/src/common/utils/useDebounce.ts:
```diff
@@ -0,0 +1,24 @@
+
+function useDebounce<T>({ value, delay }: debounceProps<T>): T {
+  const [debouncedValue, setDebouncedValue] = useState(value);
+
```
@jinhyukoo useDebounce를 범용적으로 만드는 것은 어떨까요?? 현재 useDebounce는 외부로부터 상태값을 전달받고, 해당 상태 값을 초기값으로 하는 debounceValue를 다시 상태값으로 지정해서 useEffect를 활용해 debounceValue의 변경사항을 감지하는 방식으로 보입니다! 이 방식은 디바운싱 적용이 value값에 의존하고, 컴포넌트에만 활용할 수 있어보여서요!

아래 코드는 일반적인 디바운싱 함수로 진혁님이 참고하신 사이트의 댓글에 다른분이 제안하신 코드에 네이밍만 useDebounce로 변경한 코드입니다!
```
export const  useDebounce = (func, wait) => {
  let timeout;

  return function(...args) {
    const context = this;

    if (timeout) clearTimeout(timeout);

    timeout = setTimeout(() => {
      timeout = null;
      func.apply(context, args);
    }, wait);
  };
};
```
```
const onEmailChangeListener = useDebounce(event => {
  //이메일 변경작업
}, 500);
```
위 방식을 활용하면 디바운싱 함수의 인자로 함수를 전달받아 활용가능하고, 디바운싱 작업을 컴포넌트 뿐만 아니라 다른 상황에서도 적용이 가능해보입니다!

jinhyukoo replied:
안 그래도 어제 작업 하면서 이 댓글을 봤습니당 ㅎㅎ 말씀해주신 것처럼 댓글에 나와 있는 내용이 제가 처음에 구상한 설계와 더 유사해서 위의 코드를 프로젝트에 적용시키고 싶었는데 코드가 정확히 이해가 가지 않아서 다른 방법을 사용했었어요...ㅎ 피드백 남겨주신 김에 제대로 공부해서 한 번 적용시켜보겠습니다!

wooojini commented on client/src/components/UI/molecules/LoginModalContent/LoginModalContent.tsx:
```diff
@@ -21,15 +22,60 @@ const useStyles = makeStyles((theme) => ({
+  };
+  const onPasswordChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPassword(e.target.value);
+  };
```
@jinhyukoo handler가 아닌 Listener로 네이밍을 통일하면 좋을 것 같아요 👍

jinhyukoo replied:
아이고 매번 생각없이 코딩 하다가 까먹네용 ㅋㅋㅋㅋ 수정하겠습니다!

wooojini commented on client/src/components/UI/molecules/SignUpModalContent/SignUpModalContent.tsx:
```diff
@@ -0,0 +1,92 @@
+  };
+  const onPasswordChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPassword(e.target.value);
+  };
```
@jinhyukoo 요기도 Listener로 통일하면 좋을 것 같아요! ㅎㅎ

wooojini commented on client/src/components/UI/organisms/ModalPopup/ModalPopup.tsx:
```diff
@@ -1,14 +1,15 @@
+type ModalType = TimeTableModalType | LoginModalType | SignUpModalType;
 
-const modalTypes = { ...TimeTableModalType, ...LoginModalType };
+const modalTypes = { ...TimeTableModalType, ...LoginModalType, ...SignUpModalType };
```
👍

wooojini reviewed (approved):
@jinhyukoo 고생하셨습니다! 이제 드디어 메인페이지 작업이 끝나가는군요!

jinhyukoo committed:
refactor: 네이밍 통일

jinhyukoo committed:
refactor: 네이밍 통일

wooojini committed:
refactor: SelectMenu, TimeSelectMenu 컴포넌트 코드 오타 수정

- React.useState를 useState로 수정
- TimeSelctMenu 컴포넌트 상태 변수를 selectXXX에서 selectedXXX로
  명확하게 네이밍 변경

wooojini committed:
refactor: LectureSearchFilter, TimeTableAddFormContent width 값 상수 처리

- 본문 내용 (여기에 작성)

wooojini committed:
style: Header가 스크롤 여부에 따라 움직이는 현상 제거

- 화면사이즈가 작은 경우 TimeSelectMenu가 보여질 공간이 충분치 않아서  min-height를 임시로 1000px로 설정
- Header 컴포넌트가 width 값을 가지지 않고, left, right 값으로 위치를
  조정하여 스크롤 여부에 따라 영향을 받는 문제를 width 값을 100%로 수정
  및 top 값에만 영향을 받도록 코드 수정

wooojini committed:
feat: scroll util 추가

- 브라우저의 스크롤을 조작하는 util 함수를 정의

wooojini committed:
fix: SelectMenu, TimeSelectMenu 컴포넌트 클릭시 메뉴의 위치가 올라가는 버그 수정

- 메인페이지 min-height 지정으로 인한 스크롤을 조작하여 메뉴 클릭시
  스크롤을 맨 밑으로 스크롤하여 메뉴창을 보여줄 height 확보 후, 메뉴창이
  열리도록 수정

wooojini committed:
fix: 헤더 스크롤 여부에 따른 padding 설정으로 인한 UI 움직이는 버그 수정

- SelectMenu, TimeSelectMenu 클릭 시 스크롤을 제거하고, padding-right
  값이 지정되어 UI가 움직이는 현상을 헤더에 padding-right값을 명시적으로
  지정하여 해결

wooojini committed:
Merge pull request #52 from koreatech/feat/35-select-menu

SelectMenu 컴포넌트 제작

wooojini committed:
refactor: 스토리북 코드 컨벤션 수정

- 스토리북 컴포넌트의 스토리가 하나 뿐인 경우 , Default로 네이밍
- 스토리북 컴포넌트의 스토리가 여러개 일 경우, 스토리 잘 설명해 줄 수
  있는 Type으로 간결하게 정의

wooojini committed:
Merge pull request #56 from koreatech/feat/53-storybook

스토리북 코드 컨벤션 수정

close #53

wooojini committed:
refactor: eslint 불필요한 옵션 off

- 컴포넌트 파일에 있는 eslint 옵션 off 코드 제거
- eslint 파일에 불필요한 옵션 off하는 코드 추가

wooojini committed:
Merge pull request #57 from koreatech/feat/54-eslint

eslint 불필요한 옵션 off 작업

close #54

jinhyukoo committed:
feat: useDebounce 리팩토링

jinhyukoo committed:
feat: useDebounce 리팩토링

jinhyukoo committed:
Merge branch 'feat/signup#50' of https://github.com/koreatech/hanpyo into feat/signup#50
